### PR TITLE
added basic crud for application periods

### LINKED ISSUE #3: Application period CRUD endpoints on backend
Endpoints:

GET /application-periods
- This should return all the periods and it should be pageable. (Normal users read only one, most of the time)
- The list returned needs to be sorted by `createdAt`
- For pagination it should accept a `page` parameter for the current page index (starting from 0) and a `page_size` parameter for the number of items on each page

GET /application-periods/current
- Returns the current active period `(applicationPeriodStart <= now && applicationPeriodEnd >= now)`
- It somewhat collides with the next endpoint, but all the periodId-s are numbers, so it shouldn't be a problem

GET /application-periods/{periodId}

POST /application-periods
- Creates a new application period
- The backend needs to check if the current user has permission to do it
- Application periods should not overlap

DELETE /application-periods/{periodId}
- Deletes an existing application period
- The backend needs to check if the current user has permission to do it

PATCH /application-periods/{periodId}
- Modifies an existing application period
- The backend needs to check if the current user has permission to do it
- Application periods should not overlap

### PR DESCRIPTION
**Pull request by FearsomeRover**
Closes #3 
Sry, dunno, why I thought it was #4, so the branch name is wrong.
It seems to be quite painful to fetch all application periods for checking the overlaps. Maybe we can reconsider this.

### CODE REVIEW

File: apps/backend/src/application-period/application-period.service.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    const periods = await this.prisma.applicationPeriod.findMany();
+    const newPeriodStart = new Date(createApplicationPeriodDto.applicationPeriodStartAt);
+    const newPeriodEnd = new Date(createApplicationPeriodDto.applicationPeriodEndAt);
+    periods.forEach((period) => {
+      const start = new Date(period.applicationPeriodStartAt);
+      const end = new Date(period.applicationPeriodEndAt);
+      if (start <= newPeriodEnd && end >= newPeriodStart) {
+        throw new BadRequestException('Application period overlaps with another period');
+      }
+    });
```
Comment (Tschonti) [inline apps/backend/src/application-period/application-period.service.ts:58]: possibly better solution: https://github.com/Tschonti/pontozo/blob/master/packages/functions/src/functions/seasons/create.ts#L21-L30

(elements in the where array are connected with the OR operator)

File: apps/backend/src/application-period/application-period.controller.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  @UseGuards(AuthGuard('jwt'), RolesGuard)
+  @ApiBearerAuth()
+  @Roles(Role.BODY_ADMIN, Role.BODY_MEMBER)
+  @Patch(':id')
+  async update(
+    @Body() updateApplicationPeriodDto: UpdateApplicationPeriodDto,
+    @Param() id: string
+  ): Promise<ApplicationPeriod> {
+    return this.applicationPeriodService.update(updateApplicationPeriodDto, Number(id));
+  }
```
Comment (Tschonti) [inline apps/backend/src/application-period/application-period.controller.ts:62]: not sure if this was discussed with them, but maybe it'd make sense if only the body admin could do these

File: apps/backend/src/application-period/application-period.controller.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  @Get('current')
+  async getCurrentPeriod(): Promise<ApplicationPeriod> {
+    return this.applicationPeriodService.getCurrentPeriod();
+  }
+
+  @Get(':id')
+  async findOne(@Param('id') id: string): Promise<ApplicationPeriod> {
+    return this.applicationPeriodService.findOne(Number(id));
+  }
```
Comment (Tschonti) [inline apps/backend/src/application-period/application-period.controller.ts:32]: do regular users need access to the findAll and findOne methods?

File: apps/backend/src/application-period/application-period.service.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  async findOne(id: number): Promise<ApplicationPeriod> {
+    const period = await this.prisma.applicationPeriod.findUnique({
+      where: {
+        id,
+      },
+    });
+    if (!period) {
+      throw new BadRequestException('Application period not found');
```
Comment (Tschonti) [inline apps/backend/src/application-period/application-period.service.ts:43]: that's usually a NotFoundException

File: apps/backend/src/application-period/application-period.service.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      const newPeriodStart = new Date(updateApplicationPeriodDto.applicationPeriodStartAt);
+      const newPeriodEnd = new Date(updateApplicationPeriodDto.applicationPeriodEndAt);
+      periods.forEach((period) => {
+        const start = new Date(period.applicationPeriodStartAt);
+        const end = new Date(period.applicationPeriodEndAt);
+        if (period.id !== id && start <= newPeriodEnd && end >= newPeriodStart) {
+          throw new BadRequestException('Application period overlaps with another period');
+        }
+      });
+    }
```
Comment (Tschonti) [inline apps/backend/src/application-period/application-period.service.ts:98]: same thing

File: apps/backend/src/application-period/application-period.service.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        applicationPeriodStartAt: {
+          lte: new Date(),
+        },
+        applicationPeriodEndAt: {
+          gte: new Date(),
+        },
+      },
+    });
+    if (!period) {
+      throw new Error('No current application period found');
```
Comment (Tschonti) [inline apps/backend/src/application-period/application-period.service.ts:31]: this shows up as a 500 error for the user, should be NotFoundException

**Review by Tschonti — CHANGES_REQUESTED** [review]

File: apps/backend/src/application-period/application-period.service.ts (hunk 1 above)
Comment (Tschonti) [reply apps/backend/src/application-period/application-period.service.ts]: wow yours is a much better solution xd

**Review by Tschonti — APPROVED** [review]

**Comment by Dkrisztan** [comment]
On the create-application dto  I don't see any validation, aren't those needed?

**Pull request merged**